OpenSearch (#2580)
18F/cg-site: _kbarticles/2024-12-10-using-opensearch-dashboards.md

> m ---

Reviewer: ```suggestion
---
```

I wonder if this stray character is messing up the front matter and redirect

Author: Yes, that was it. That typo snuck in between commits, really confusing me....

layout: post
title: "Understanding Opensearch Dashboards, logging and how to visualize your application logs"
date: December 10, 2024
excerpt: This article briefly explains what opensearch dashboards does, what types of logs are available in Dashboards, creating dashboards, and how to create log visualizations in Dashboards.
redirect_from:
  - /knowledge-base/2022-12-08-common-kibana-searches-on-clouddotgov/

---

## What does Dashboard do?

Dashboard is a user interface that lets you search and visualize your application logs. Dashboard has a [user guide](https://opensearch.org/docs/latest/dashboards/) that explains more about how to use it and to create custom visualizations.

## What are some types of logs?

> Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type. HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

Reviewer: ```suggestion
Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type. HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

The log type is stored on logs in the `@source.type` field. So to query for application logs, you could use a filter of `@source.type: "APP"`.
```

> Cloudfoundry logs are often translated into field names in Opensearch/Dashboards using the log type as a prefix. For example, APP logs in Dashboards include fields like `app.name` for the application name.

Reviewer: This isn't true. I think it's referring to how JSON logs were parsed into Elasticsearch.

If a JSON log comes in now and contains a key of `name`, then you could search `app.name`, but it all works very differently and there is no creation of custom fields

## Querying Logs

Cloudfoundry logs have a few useful fields which will be helpful for querying logs.
Cloudfoundry has 3 fields that can be very helpful to filter down what app logs show up.

> @cf.org - The organization, this is applicable to multi-org users
> @cf.space - The space in a organization.
> @cf.app - The app in a space.

Reviewer: ```suggestion

- `@cf.org` - The organization, this is applicable to multi-org users
- `@cf.space` - The space in a organization.
- `@cf.app` - The app in a space.
```

> For further query
> @message - the app specific message attached to the log
> @raw - the entire log, useful when looking for cloudfoundry specific values, when looking for app use @message

Reviewer: ```suggestion
Other fields that may be useful for querying:

`@message` - the app specific message attached to the log. This field supports full-text searching, so you can do a search of `@message: "foo"` to find all logs where `@message` contains `foo`. 
`@raw` - the raw message received by OpenSearch before parsing into other fields. This field also supports full-text searching
```

- **rtr.hostname**: The domain/hostname the request was sent to (e.g test.app.cloud.gov)
- **rtr.http_user_agent**: What user agent the request came from (Chrome, Firefox, Curl, etc…)
- **rtr.path**: The specific url path that was requested (e.g. /my/test/page)
- **rtr.status**: Gives the status of the request (200, 404, etc…)
- **rtr.verb**: The type of request (POST, GET, etc...)
- **rtr.x_forwarded_for**: The IP address the request came from
- **rtr.timestamp**: The time of the request in UTC

The full list of fields available for router logs can be found in [our Opensearch field mapping configuration](https://github.com/cloud-gov/opensearch-boshrelease/blob/5100229e391237f4a41df023338a247c55276ee8/jobs/opensearch_templates/templates/component-index-mappings-app.json.erb#L44).

## Generate report of logs.

Opensearch allows for [generating a report](https://opensearch.org/docs/latest/reporting/report-dashboard-index/) from a saved search. This is done by choosing a search and clicking the **Reporting** menu option.

![Screenshot of Dashboards interface showing the selection of ]({{ site.baseurl }}/assets/images/content/opensearch-reporting.png)

> > Reports have a non-configurable 10,000 row limit. They have no explicit size limit (for example, MB),

Reviewer: The formatting on block quotes doesn't look great. Maybe we remove the `>` and just prepend this section with `Please note:` in bold

> > but extremely large documents could cause report generation to fail with an out of memory error from the V8 JavaScript engine

Reviewer: ```suggestion
> but extremely large documents could cause report generation to fail
```

I don't think the error would come from Javascript/V8

---
layout: post
title: "Understanding OpenSearch Dashboards, logging and how to visualize your application logs"
date: December 10, 2024
excerpt: This article briefly explains what opensearch dashboards does, what types of logs are available in Dashboards, creating dashboards, and how to create log visualizations in Dashboards.
redirect_from:
  - /knowledge-base/2022-12-08-common-kibana-searches-on-clouddotgov/

---

> ## What does Dashboard do?

Reviewer: ```suggestion
## What does Dashboards do?
```

Technically, the name is OpenSearch Dashboards

> Dashboard is a user interface that lets you search and visualize your application logs. Dashboard has a [user guide](https://opensearch.org/docs/latest/dashboards/) that explains more about how to use it and to create custom visualizations.

Reviewer: ```suggestion
Dashboards is a user interface that lets you search and visualize your application logs. Dashboards has a [user guide](https://opensearch.org/docs/latest/dashboards/) that explains more about how to use it and to create custom visualizations.
```

The log type is stored on logs in the `@source.type` field. So to query for application logs, you could use a filter of `@source.type: "APP"`.

## Querying Logs

Cloud Foundry logs have a few useful fields which will be helpful for querying logs.
Cloud Foundry has 3 fields that can be very helpful to filter down what app logs show up.

- `@cf.org` - The organization, this is applicable to multi-org users
- `@cf.space` - The space in a organization.
- `@cf.app` - The app in a space.

Other fields that may be useful for querying:

- `@message` - the app specific message attached to the log. This field supports full-text searching, so you can do a search of `@message: "foo"` to find all logs where `@message` contains `foo`. 
- `@raw` - the raw message received by OpenSearch before parsing into other fields. This field also supports full-text searching

## How to visualize application traffic

Router log data can be used to create a visualization of your application traffic following the steps below.

> After you have logged into Dashboard, click "Discover" in the left sidebar menu. Then, add filters(under search bar) and search terms to query for router logs as seen in the screenshot below. Please note that the filters shown here for a specific space and application are just an example. You might want to view logs for all requests to application in a given space, in which case you would not want a filter for `@cf.app`.

Reviewer: ```suggestion
After you have logged into Dashboards, click "Discover" in the left sidebar menu. Then, add filters(under search bar) and search terms to query for router logs as seen in the screenshot below. Please note that the filters shown here for a specific space and application are just an example. You might want to view logs for all requests to application in a given space, in which case you would not want a filter for `@cf.app`.
```

![Screenshot of Dashboards visualization interface showing a line chart of requests over time with a panel for chart configuration options visible on the right side of the screen. In the chart configuration options, "line" is selected as the chart type]({{ site.baseurl }}/assets/images/content/dashboard-line.png)

## Fields for router requests

Listed below are the explanations of some field names for router (RTR) logs:

- **rtr.app.id**: The application guid
- **rtr.hostname**: The domain/hostname the request was sent to (e.g test.app.cloud.gov)
- **rtr.http_user_agent**: What user agent the request came from (Chrome, Firefox, Curl, etc…)
- **rtr.path**: The specific url path that was requested (e.g. /my/test/page)
- **rtr.status**: Gives the status of the request (200, 404, etc…)
- **rtr.verb**: The type of request (POST, GET, etc...)
- **rtr.x_forwarded_for**: The IP address the request came from
- **rtr.timestamp**: The time of the request in UTC

> The full list of fields available for router logs can be found in [our OpenSearch field mapping configuration](https://github.com/cloud-gov/opensearch-boshrelease/blob/main/jobs/opensearch_templates/templates/component-index-mappings-app.json.erb#L46)

Reviewer: ```suggestion
The full list of fields available for router logs can be found in [our OpenSearch field mapping configuration](https://github.com/cloud-gov/opensearch-boshrelease/blob/2078ff06dec204c12bc64a45fe707f71e1c9e303/jobs/opensearch_templates/templates/component-index-mappings-app.json.erb#L46)
```

You can link to a commit permalink, so that the reference will always stay valid. In the future, if we want to update the link, we can create a new permalink for the latest commit

---
layout: post
title: "Understanding OpenSearch Dashboards, logging and how to visualize your application logs"
date: December 10, 2024
excerpt: This article briefly explains what opensearch dashboards does, what types of logs are available in Dashboards, creating dashboards, and how to create log visualizations in Dashboards.
redirect_from:
  - /knowledge-base/2022-12-08-common-kibana-searches-on-clouddotgov/

---

## What does Dashboard do?

Dashboard is a user interface that lets you search and visualize your application logs. Dashboard has a [user guide](https://opensearch.org/docs/latest/dashboards/) that explains more about how to use it and to create custom visualizations.

## What are some types of logs?

> Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type. HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

Reviewer: ```suggestion
Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type. HTTP requests being routed to an app will produce the RTR log type. The various types of logs [are listed in the CloudFoundry documentation](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).
```

The log type is stored on logs in the `@source.type` field. So to query for application logs, you could use a filter of `@source.type: "APP"`.

> Cloud Foundry logs are often translated into field names in OpenSearch Dashboards using the log type as a prefix. For example, APP logs in Dashboards include fields like `app.name` for the application name.

Reviewer: This isn't true. It seems to refer to how JSON logs were parsed into Kibana, which is different with flat_object in OpenSearch. No custom field names are created from logs in OpenSearch

Author: Line 21 -- this was carried over from the Kibana doc, and I don't know that it adds any information, so deleting.

## Querying Logs

Cloud Foundry logs have a few useful fields which will be helpful for querying logs.
Cloud Foundry has 3 fields that can be very helpful to filter down what app logs show up.

- `@cf.org` - The organization, this is applicable to multi-org users
- `@cf.space` - The space in a organization.
- `@cf.app` - The app in a space.

Other fields that may be useful for querying:

`@message` - the app specific message attached to the log. This field supports full-text searching, so you can do a search of `@message: "foo"` to find all logs where `@message` contains `foo`. 

> `@raw` - the raw message received by OpenSearch before parsing into other fields. This field also supports full-text searching

Reviewer: ```suggestion
- `@message` - the app specific message attached to the log. This field supports full-text searching, so you can do a search of `@message: "foo"` to find all logs where `@message` contains `foo`. 
- `@raw` - the raw message received by OpenSearch before parsing into other fields. This field also supports full-text searching
```
